test/example.c

```diff
@@ -19,13 +19,76 @@
 #  define TESTFILE "foo.gz"
 #endif
 
-#define CHECK_ERR(err, msg) { \
+#define SUCCESSFUL                 1
+#define FAILED_WITH_ERROR_CODE     0
+#define FAILED_WITHOUT_ERROR_CODE -1
+
+typedef struct test_result_s {
+    int           result; /* One of: SUCCESSFUL,
+                          FAILED_WITH_ERROR_CODE, or
+                          FAILED_WITHOUT_ERROR_CODE*/
+    int           err;     /* error code if success is FAILED_WITH_ERROR_CODE */
+    z_const char* message;
+    z_const char* extended_message;
+} test_result;
+
+// TODO(cblume): some of these can be functions instead of macros
+#define EXIT_ON_ERROR(err, msg) { \
     if (err != Z_OK) { \
         fprintf(stderr, "%s error: %d\n", msg, err); \
         exit(1); \
     } \
 }
 
+#define RETURN_ON_ERROR_WITH_MESSAGE(err, msg, result) { \
+    if (err != Z_OK) { \
+        result.err = err; \
+        result.result = FAILED_WITH_ERROR_CODE; \
+        result.message = msg; \
+        return result; \
+    } \
+}
+
+#define RETURN_WITH_MESSAGE(msg, result) { \
+    result.result = FAILED_WITHOUT_ERROR_CODE; \
+    result.message = msg; \
+    return result; \
+}
+
+#define HANDLE_TEST_RESULTS(output, result, testcase_name, is_junit_output) { \
+    if (is_junit_output) { \
+        fprintf(output, "\t\t<testcase name=\"%s\">", testcase_name); \
+    } \
+    if (result.result == FAILED_WITH_ERROR_CODE) { \
+        if (is_junit_output) { \
+            fprintf(output, "\n\t\t\t<failure>%s error: %d</failure>\n\t\t", result.message, result.err); \
+		} else { \
+            fprintf(stderr, "%s error: %d\n", result.message, result.err); \
+            exit(1); \
+		} \
+    } else if (result.result == FAILED_WITHOUT_ERROR_CODE) { \
+        if (is_junit_output) { \
+            fprintf(output, "\n\t\t\t<failure>%s</failure>\n\t\t", result.message); \
+		} else { \
+            fprintf(stderr, result.message); \
```

Comment:
Might want to use "%s", result.message:

../test/example.c:707:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:711:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "deflate", is_junit_output);
     ^~~~~~~~~~~~~~~~~~~
../test/example.c:711:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:713:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "inflate", is_junit_output);
     ^~~~~~~~~~~~~~~~~~~
../test/example.c:713:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:716:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "large deflate", is_junit_output);
     ^~~~~~~~~~~~~~~~~~~
../test/example.c:716:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:718:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "large inflate", is_junit_output);